netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,313 @@
+ */
+public class BufferingHttp2ConnectionEncoder extends DefaultHttp2ConnectionEncoder {
+
+  // Buffer for any streams and corresponding frames that could not be created
+  // due to the maximum concurrent stream limit being hit.
+  private TreeMap<Integer, PendingStream> pendingStreams = new TreeMap<Integer, PendingStream>();
+  // Smallest stream id whose corresponding frames do not get buffered.
+  int largestCreatedStreamId;
+
+  protected BufferingHttp2ConnectionEncoder(Builder builder) {
+    super(builder);
+    connection().addListener(new Http2ConnectionAdapter() {
+      @Override
+      public void goingAway() {
+        cancelPendingStreams();
+      }
+
+      @Override
+      public void streamInactive(Http2Stream stream) {
+        createNextPendingStream();
+      }
+    });
+  }
+
+  /**
+   * Builder for new instances of {@link BufferingHttp2ConnectionEncoder}.
+   */
+  public static class Builder extends DefaultHttp2ConnectionEncoder.Builder {
+    @Override
+    public Builder connection(Http2Connection connection) {
+      this.connection = connection;
+      return this;
+    }
+
+    @Override
+    public Builder lifecycleManager(Http2LifecycleManager lifecycleManager) {
+      this.lifecycleManager = lifecycleManager;
+      return this;
+    }
+
+    @Override
+    public Http2LifecycleManager lifecycleManager() {
+      return lifecycleManager;
+    }
+
+    @Override
+    public Builder frameWriter(Http2FrameWriter frameWriter) {
+      this.frameWriter = frameWriter;
+      return this;
+    }
+
+    @Override
+    public Http2ConnectionEncoder build() {
+      return new BufferingHttp2ConnectionEncoder(this);
+    }
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  @Override
+  public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                                    int padding, boolean endStream, ChannelPromise promise) {
+    return writeHeaders(ctx, streamId, headers, 0, Http2CodecUtil.DEFAULT_PRIORITY_WEIGHT, false,
+                        padding, endStream, promise);
+  }
+
+  @Override
+  public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                                    int streamDependency, short weight, boolean exclusive,
+                                    int padding, boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeHeaders(ctx, streamId, headers, streamDependency, weight, exclusive,
+                                padding, endOfStream, promise);
+    }
+    if (connection().local().acceptingNewStreams()) {
+      assert streamId > largestCreatedStreamId;
+      largestCreatedStreamId = streamId;
+      return super.writeHeaders(ctx, streamId, headers, streamDependency, weight, exclusive,
+                                padding, endOfStream, promise);
+    }
+    if (connection().isGoAway()) {
```

Comment:
You could probably also just called the super method in this case.  Maybe change the first if to `if (existingStream(streamId) || connection().isGoAway())`